code/game/objects/structures/window.dm

```diff
@@ -180,50 +180,65 @@
 		return
 	..()
 
-/obj/structure/window/attackby(obj/item/I, mob/living/user, params)
+/obj/structure/window/tool_act(mob/living/user, obj/item/tool, tool_type, is_right_clicking)
 	if(!can_be_reached(user))
 		return TRUE //skip the afterattack
-
 	add_fingerprint(user)
+	return ..()
 
-	if(I.tool_behaviour == TOOL_WELDER)
-		if(atom_integrity < max_integrity)
-			if(!I.tool_start_check(user, amount = 0))
-				return
+/obj/structure/window/welder_act(mob/living/user, obj/item/tool)
+	if(atom_integrity < max_integrity)
+		if(!tool.tool_start_check(user, amount = 0))
+			return FALSE
 
-			to_chat(user, span_notice("You begin repairing [src]..."))
-			if(I.use_tool(src, user, 40, volume = 50))
-				atom_integrity = max_integrity
-				update_nearby_icons()
-				to_chat(user, span_notice("You repair [src]."))
-		else
-			to_chat(user, span_warning("[src] is already in good condition!"))
-		return
+		to_chat(user, span_notice("You begin repairing [src]..."))
+		if(tool.use_tool(src, user, 4 SECONDS, volume = 50))
+			atom_integrity = max_integrity
+			update_nearby_icons()
+			to_chat(user, span_notice("You repair [src]."))
+	else
+		to_chat(user, span_warning("[src] is already in good condition!"))
+	return TRUE
```

Comment:
early returns here? a good pattern is to do the least complex check option first. Helps the readability of the code since people don't need to remember exactly what that else is elsing